File: windows/tests/ConnectorControl.Core.Tests/State/AppStateUpgradeTests.cs
using ConnectorControl.Core.State;
using ConnectorControl.Core.Tests.TestSupport;

namespace ConnectorControl.Core.Tests.State;

/// <summary>
/// Mirror: Tests/ConnectorControlStateTests/AppStateUpgradeTests.swift
///
/// An install upgraded from 1.3, which knew collections as profiles: Tests/Fixtures/v1.3/ holds the
/// master list 1.3 wrote (schema 2, two profiles, "Work" active, some connectors off) and the
/// Claude config it left applied. There is no collections.json and no collections-local.json: 1.3
/// never wrote either.
/// </summary>
public class AppStateUpgradeTests
{
    /// <summary>Each profile's connectors and whether they are on, as 1.3 wrote them.</summary>
    private static readonly Dictionary<string, Dictionary<string, bool>> Flags13 = new(StringComparer.Ordinal)
    {
        ["Default"] = new(StringComparer.Ordinal) { ["aws-mcp"] = true, ["scoutbook"] = false, ["service-now"] = true },
        ["Work"] = new(StringComparer.Ordinal) { ["jira"] = true, ["ledger"] = false, ["scoutbook"] = true },
    };

    /// <summary>Places the 1.3 install's two files where this app looks for them.</summary>
    private static (byte[] Store, byte[] Claude) InstallVersion13(AppStateHarness h)
    {
        var store = Fixtures.Bytes(Path.Combine("v1.3", "mcps.json"));
        var claude = Fixtures.Bytes(Path.Combine("v1.3", "claude_desktop_config.json"));
        Directory.CreateDirectory(h.StoreDir);
        File.WriteAllBytes(h.MasterStorePath, store);
        File.WriteAllBytes(h.ClaudeConfigPath, claude);
        return (store, claude);
    }

    /// <summary>The keys 1.3's decoder requires at every level above a connector's own config: the
    /// file's root, each profile, and each connector entry. 1.3 fails to decode a file that lacks
    /// one.</summary>
    private static List<string> Shape(byte[] data)
    {
        static string Keys(JsonValue value, string path) => $"{path}: {string.Join(",", value.ObjectProperties.Keys)}";
        var root = JsonValue.Parse(data);
        var shape = new List<string> { Keys(root, "") };
        foreach (var (name, profile) in root.ObjectProperties["profiles"].ObjectProperties)
        {
            shape.Add(Keys(profile, $"/profiles/{name}"));
            foreach (var (connector, entry) in profile.ObjectProperties["mcps"].ObjectProperties)
            {
                shape.Add(Keys(entry, $"/profiles/{name}/mcps/{connector}"));
            }
        }
        return shape;
    }

    [Fact]
    public void AVersion13InstallOpensItsProfilesAsLocalCollectionsAndStaysReadableBy13()
    {
        using var h = new AppStateHarness(seedClaudeConfig: false);
        var written = InstallVersion13(h);
        Assert.False(File.Exists(Path.Combine(h.StoreDir, CollectionsFile.FileName)));
        Assert.False(File.Exists(Path.Combine(h.StoreDir, CollectionsLocalCache.FileName)));

        using var state = h.Create();

        // Every profile is a local collection under its own name, and the active one stays active.
        Assert.Equal(["Default", "Work"], state.CollectionNames);
        foreach (var name in state.CollectionNames)
        {
            Assert.Equal(CollectionKind.Local, state.KindOf(name));
            Assert.False(state.IsPublished(name), name);
        }
        Assert.Equal("Work", state.ActiveCollection);
        Assert.Null(state.CollectionBanner);
        Assert.Null(state.LastError);

        // Connectors, their configs and their on/off flags are exactly what 1.3 wrote.
        Assert.Equal(MasterStoreIO.Read(Fixtures.Path(Path.Combine("v1.3", "mcps.json"))), state.Store);
        Assert.Equal(Flags13, state.Store.Collections.ToDictionary(
            c => c.Key, c => c.Value.Mcps.ToDictionary(m => m.Key, m => m.Value.Enabled, StringComparer.Ordinal), StringComparer.Ordinal));

        // The upgrade alone gives Claude nothing new to run, so its config is left as it was.
        Assert.Equal(written.Claude, File.ReadAllBytes(h.ClaudeConfigPath));
        Assert.Null(h.Settings.LastApplyDate);   // nothing was applied
        Assert.False(state.NeedsClaudeRestart);

        // The master list keeps 1.3's keys, whether or not the launch rewrote it...
        var shape13 = Shape(written.Store);
        Assert.Equal(shape13, Shape(File.ReadAllBytes(h.MasterStorePath)));
        Assert.Equal(state.Store, h.StoreOnDisk());

        // ...and after a save from this version, which a 1.3 install on another machine or after a
        // downgrade still has to read: the same keys, schema version 2, the same content.
        state.SetEnabled("ledger", true, "Work");
        state.SetEnabled("ledger", false, "Work");
        var saved = File.ReadAllBytes(h.MasterStorePath);
        Assert.Equal(shape13, Shape(saved));
        Assert.Equal(JsonValue.Int(2), JsonValue.Parse(saved).ObjectProperties["version"]);
        Assert.Equal(JsonValue.Parse(written.Store), JsonValue.Parse(saved));

        // With nothing published or subscribed, this launch left the sidecar valid and empty.
        // The next launch reads it as empty, not as unreadable, so the first publish still lands.
        var sidecar = Path.Combine(h.StoreDir, CollectionsFile.FileName);
        Assert.Equal(new CollectionsFile([]), CollectionsFile.Decode(JsonValue.Parse(File.ReadAllBytes(sidecar))));
        state.Dispose();
        using var second = h.Create();
        h.Publish(second, "Work");
        Assert.Null(second.LastError);
        Assert.NotNull(CollectionsFile.Load(sidecar).Collections["Work"].Publish);
    }
}

File: Tests/ConnectorControlStateTests/AppStateUpgradeTests.swift
import XCTest
import ConnectorControlCore
import ConnectorControlTestSupport
@testable import ConnectorControlState

/// Mirror: windows/tests/ConnectorControl.Core.Tests/State/AppStateUpgradeTests.cs
///
/// An install upgraded from 1.3, which knew collections as profiles: `Tests/Fixtures/v1.3/` holds
/// the master list 1.3 wrote (schema 2, two profiles, "Work" active, some connectors off) and the
/// Claude config it left applied. There is no collections.json and no collections-local.json:
/// 1.3 never wrote either.
@MainActor
final class AppStateUpgradeTests: XCTestCase {
    /// Each profile's connectors and whether they are on, as 1.3 wrote them.
    private let flags13: [String: [String: Bool]] = [
        "Default": ["aws-mcp": true, "scoutbook": false, "service-now": true],
        "Work": ["jira": true, "ledger": false, "scoutbook": true],
    ]

    /// Places the 1.3 install's two files where this app looks for them.
    private func installVersion13(_ h: AppStateHarness) throws -> (store: Data, claude: Data) {
        let store = try Data(contentsOf: Fixtures.url("v1.3/mcps.json"))
        let claude = try Data(contentsOf: Fixtures.url("v1.3/claude_desktop_config.json"))
        try FileManager.default.createDirectory(at: h.storeDir, withIntermediateDirectories: true)
        try store.write(to: h.masterStoreURL)
        try claude.write(to: h.claudeConfigURL)
        return (store, claude)
    }

    /// The keys 1.3's decoder requires at every level above a connector's own config: the file's
    /// root, each profile, and each connector entry. 1.3 fails to decode a file that lacks one.
    private func shape(of data: Data) throws -> [String] {
        func keys(_ value: JSONValue?, _ path: String) throws -> (String, [String: JSONValue]) {
            guard case .object(let object)? = value else { throw AppStateHarness.HarnessError() }
            return ("\(path): \(object.keys.sorted().joined(separator: ","))", object)
        }
        let (root, rootObject) = try keys(try JSONValue.parse(data), "")
        var shape = [root]
        guard case .object(let profiles)? = rootObject["profiles"] else { throw AppStateHarness.HarnessError() }
        for name in profiles.keys.sorted() {
            let (line, profile) = try keys(profiles[name], "/profiles/\(name)")
            shape.append(line)
            guard case .object(let mcps)? = profile["mcps"] else { throw AppStateHarness.HarnessError() }
            for connector in mcps.keys.sorted() {
                shape.append(try keys(mcps[connector], "/profiles/\(name)/mcps/\(connector)").0)
            }
        }
        return shape
    }

    func testAVersion13InstallOpensItsProfilesAsLocalCollectionsAndStaysReadableBy13() throws {
        let h = AppStateHarness(seedClaudeConfig: false)
        defer { h.dispose() }
        let written = try installVersion13(h)
        let fm = FileManager.default
        XCTAssertFalse(fm.fileExists(atPath: h.storeDir.appendingPathComponent(CollectionsFile.fileName).path))
        XCTAssertFalse(fm.fileExists(atPath: h.storeDir.appendingPathComponent(CollectionsLocalCache.fileName).path))

        let state = h.create()

        // Every profile is a local collection under its own name, and the active one stays active.
        XCTAssertEqual(state.collectionNames, ["Default", "Work"])
        for name in state.collectionNames {
            XCTAssertEqual(state.kind(of: name), .local, name)
            XCTAssertFalse(state.isPublished(name), name)
        }
        XCTAssertEqual(state.activeCollection, "Work")
        XCTAssertNil(state.collectionBanner)
        XCTAssertNil(state.lastError)

        // Connectors, their configs and their on/off flags are exactly what 1.3 wrote.
        XCTAssertEqual(state.store, MasterStoreIO.read(from: Fixtures.url("v1.3/mcps.json")))
        XCTAssertEqual(state.store.collections.mapValues { $0.mcps.mapValues(\.enabled) }, flags13)

        // The upgrade alone gives Claude nothing new to run, so its config is left as it was.
        XCTAssertEqual(try Data(contentsOf: h.claudeConfigURL), written.claude)
        XCTAssertNil(h.settings.lastApplyDate, "nothing was applied")
        XCTAssertFalse(state.needsClaudeRestart)

        // The master list keeps 1.3's keys, whether or not the launch rewrote it...
        let shape13 = try shape(of: written.store)
        XCTAssertEqual(try shape(of: Data(contentsOf: h.masterStoreURL)), shape13)
        XCTAssertEqual(try h.storeOnDisk(), state.store)

        // ...and after a save from this version, which a 1.3 install on another machine or after
        // a downgrade still has to read: the same keys, schema version 2, the same content.
        state.setEnabled("ledger", true, in: "Work")
        state.setEnabled("ledger", false, in: "Work")
        let saved = try Data(contentsOf: h.masterStoreURL)
        XCTAssertEqual(try shape(of: saved), shape13)
        XCTAssertEqual(try JSONValue.parse(saved).value(at: JSONPointer(["version"])), .int(2))
        XCTAssertEqual(try JSONValue.parse(saved), try JSONValue.parse(written.store))
    }
}
